flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperator.java

```diff
@@ -94,20 +102,38 @@
 	/** Backend for keyed state. This might be empty if we're not on a keyed stream. */
 	private transient KeyedStateBackend<?> keyedStateBackend;
 
-	protected transient MetricGroup metrics;
+	// --------------- Metrics ---------------------------
+
+	/** Metric group for the operator */
+	protected MetricGroup metrics;
+
+	/** Flag indicating if this operator is a sink */
+	protected transient boolean isSink = false;
+
+	protected LatencyGauge latencyGauge;
+
 
 	// ------------------------------------------------------------------------
 	//  Life Cycle
 	// ------------------------------------------------------------------------
 
 	@Override
-	public void setup(StreamTask<?, ?> containingTask, StreamConfig config, Output<StreamRecord<OUT>> output) {
+	public void setup(StreamTask<?, ?> containingTask, StreamConfig config, Output<StreamRecord<OUT>> output, boolean isSink) {
 		this.container = containingTask;
 		this.config = config;
 		String operatorName = containingTask.getEnvironment().getTaskInfo().getTaskName().split("->")[config.getChainIndex()].trim();
 		
 		this.metrics = container.getEnvironment().getMetricGroup().addOperator(operatorName);
 		this.output = new CountingOutput(output, this.metrics.counter("numRecordsOut"));
+		this.isSink = isSink;
+		Configuration taskManagerConfig = container.getEnvironment().getTaskManagerInfo().getConfiguration();
+		int historySize = taskManagerConfig.getInteger(ConfigConstants.METRICS_LATENCY_HISTORY_SIZE, ConfigConstants.DEFAULT_METRICS_LATENCY_HISTORY_SIZE);
+		if(historySize <= 0) {
+			LOG.warn("{} has been set to a value below 0: {}. Using default.", ConfigConstants.METRICS_LATENCY_HISTORY_SIZE, historySize);
+			historySize = ConfigConstants.DEFAULT_METRICS_LATENCY_HISTORY_SIZE;
+		}
+
+		latencyGauge = new LatencyGauge(this.metrics, historySize, !isSink);
```

Comment:
this looks a bit odd; `latencyGauge = this.metrics.gauge(new LatencyGauge(historySize, isSink), "latency")` would be more consistent to how other metrics are registered.